src/os/filestore/FileStore.cc

```diff
@@ -3536,17 +3536,30 @@ int FileStore::_do_sparse_copy_range(int from, int to, uint64_t srcoff, uint64_t
   if (len == 0)
     return 0;
 
-  if (backend->has_seek_data_hole()) {
+  int64_t written = 0;
+  struct stat st;
+  r = ::fstat(to, &st);
+  if (r < 0) {
+    r = -errno;
+    derr << __func__ << ": fstat error at " << to << " " << cpp_strerror(r) << dendl;
+    goto out;
+  }
+
+  if ((uint64_t)st.st_size > dstoff) {
```

Comment:
I suspect in the hole case falling back to a full copy is fine.  That path shouldn't be triggered with real workloads anyway.